types/convert_test.go

```diff
@@ -538,48 +538,73 @@ func (s *testTypeConvertSuite) TestConvert(c *C) {
 	signedAccept(c, mysql.TypeTiny, -128, "-128")
 	signedAccept(c, mysql.TypeTiny, 127, "127")
 	signedDeny(c, mysql.TypeTiny, 128, "127")
+	signedAccept(c, mysql.TypeTiny, NewBinaryLiteralFromUint(127, -1), "127")
+	signedDeny(c, mysql.TypeTiny, NewBinaryLiteralFromUint(128, -1), "128")
```

Comment:
Comparing with line 540, the result should still be `"127"` instead of `"128"`.

Reply:
Here is my consideration:
For  `select cast(0x99999999999999999999999999 as sign)`, result should be -1.
Thus I didn't truncate the 128 to 127

Reply:
Can we return int64max for `select cast(0x99999999999999999999999999 as signed)`?
This may result in inconsistent with MySQL, however, I think it's more consistent with 'select cast(99999999999999999999999999 as signed)` which return int64max.

Reply:
@wjhuang2016 My consideration is that on MySQL 8

```sql
set sql_mode = '';
create table a (a bigint signed);
insert into a values (0x99999999999999999999999999);
select * from a;
```

returns 

```
+---------------------+
| a                   |
+---------------------+
| 9223372036854775807 |
+---------------------+
1 row in set (0.00 sec)
```

Not `-1`. If the conversion behavior of `CAST` and `INSERT` differ, maybe we need to introduce a different branch for `CAST`? (e.g. check the current `StatementContext` and return different values)

Reply:
We can return `-1` for special situation. But I wouder why the `-1` is returned in MySQL, maybe it's a  bug. 
The logic that return `-1` in TiDB is:
https://github.com/pingcap/tidb/blob/9251724533e1e82d43d6b7ff5045218db0009e30/types/binary_literal.go#L104-L123

This function only consider that turn the binary literal into uint64.
When the length of  binary literal > 8, it return uint64max. 
If we try to return `-1` and check the `StatementContext`, the code would be complex.
So I suggest that we should return int64max instead of `-1`, what's your opinion?

Reply:
@wjhuang2016 MySQL isn't returning -1, but rather `uint64max`, which is then bit-cast into a signed `int64` which is `-1`.

```sql
mysql> select cast(0x999999999999999999999999999999999999 as unsigned);
18446744073709551615
mysql> select 0x999999999999999999999999999999999999 + 0;
18446744073709551615
mysql> select cast(18446744073709551615 as signed);
-1
mysql> select cast(0x8111111111111111 as unsigned);
9300233470495232273
mysql> select 0x8111111111111111 + 0;
9300233470495232273
mysql> select cast(0x8111111111111111 as signed);
-9146510603214319343
mysql> select cast(9300233470495232273 as signed);
-9146510603214319343
```

The reason `cast(99999999999999999999999999 as signed)` returns int64max is because 99999999999999999999999999 is treated as a DECIMAL instead of a BIGINT.

```sql
mysql> select cast(18446744073709551615 as signed);
-1
mysql> select cast(cast(18446744073709551615 as decimal(20)) as signed);
9223372036854775807
```

In summary:

* `CAST(x AS SIGNED)` for integers and binary literals first casts `x` into an `uint64`, and then bit-cast it as an `int64`
* `CAST(x AS SIGNED)` for real and decimal convert `x` to `int64` directly
* When `INSERT`ing the binary literal `x` into the table, it also directly cast `x` into an `int64`.

With this, I think it is fine to return `int64max` from the functions you've modified, but please include a test in `builtin_cast_test.go` to ensure the behavior `CAST(0x99999999999999999999999999 AS SIGNED)` still returned `-1` (I hope the existing TiDB code has already applied the x → uint64 → int64 detour.)

Reply:
@kennytm Then the problem is that in TiDB, `INSERT` and `CAST` use the same logic. 
`(b BinaryLiteral) ToInt` will turn the binart literal into uint64 first. It seems that we need to distinguish the return type in `ToInt`

Reply:
Besides, 

> CAST(x AS SIGNED) for integers and binary literals first casts x into an uint64, and then bit-cast it as an int64

For integer it needn't to cast the x into uint64. I think cast(x AS ASGNED) for binary literals is a exception.

Reply:
@kennytm @qw4990 PTAL

Reply:
@zz-jason The blockage here is that we have different opinion on the expected behavior, unrelated to the code quality.